connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Collections;
+import java.util.stream.Collectors;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.locks.ReentrantLock;
+import java.time.Duration;
+
+/** Emits checkpoints for upstream consumer groups. */
+public class MirrorCheckpointTask extends SourceTask {
+
+    private static final Logger log = LoggerFactory.getLogger(MirrorCheckpointTask.class);
+
+    private AdminClient sourceAdminClient;
+    private String sourceClusterAlias;
+    private String targetClusterAlias;
+    private String checkpointsTopic;
+    private Duration interval;
+    private Duration pollTimeout;
+    private TopicFilter topicFilter;
+    private Set<String> consumerGroups;
+    private ReplicationPolicy replicationPolicy;
+    private OffsetSyncStore offsetSyncStore;
+    private boolean stopped;
+    private ReentrantLock lock;
+    private MirrorMetrics metrics;
+
+    public MirrorCheckpointTask() {}
+
+    // for testing
+    MirrorCheckpointTask(String sourceClusterAlias, String targetClusterAlias,
+            ReplicationPolicy replicationPolicy, OffsetSyncStore offsetSyncStore) {
+        this.sourceClusterAlias = sourceClusterAlias;
+        this.targetClusterAlias = targetClusterAlias;
+        this.replicationPolicy = replicationPolicy;
+        this.offsetSyncStore = offsetSyncStore;
+    }
+
+    @Override
+    public void start(Map<String, String> props) {
+        lock = new ReentrantLock();
+        MirrorTaskConfig config = new MirrorTaskConfig(props);
+        stopped = false;
+        sourceClusterAlias = config.sourceClusterAlias();
+        targetClusterAlias = config.targetClusterAlias();
+        consumerGroups = config.taskConsumerGroups();
+        checkpointsTopic = config.checkpointsTopic();
+        topicFilter = config.topicFilter();
+        replicationPolicy = config.replicationPolicy();
+        interval = config.emitCheckpointsInterval();
+        pollTimeout = config.consumerPollTimeout();
+        offsetSyncStore = new OffsetSyncStore(config);
+        sourceAdminClient = AdminClient.create(config.sourceAdminConfig());
+        metrics = config.metrics();
+    }
+
+    @Override
+    public void commit() throws InterruptedException {
+        // nop
+    }
+
+    @Override
+    public void stop() {
+        stopped = true;
+        new Thread(this::cleanup).start();  // cleanup off-thread to prevent blocking
+    }
+
+    private void cleanup() {
+        lock.lock();
+        try {
+            offsetSyncStore.close();
+            sourceAdminClient.close();
```

Comment:
The underlying client method seems to use a default timeout of Long.MAX_VALUE, TimeUnit.MILLISECONDS.
I'm worried that it can prevent the JVM to stop ?

Reply:
I feel like I must be missing something obvious but since the source admin client will be an instance of `KafkaAdminClient` since we're calling `https://github.com/ryannedolan/kafka/blob/KIP-382/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java#L83` to create the client and the `KafkaAdminClient` overrides close as such https://github.com/ryannedolan/kafka/blob/KIP-382/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L468 are we not actually releasing resources but calling close without a duration? I'm probably misremembering something about inheritance but either way I think we should be passing in a duration here since the default close just is kind of odd vpernin pointed out.

Reply:
Thanks guys, this does seem like a problem. Fixed by adding a configurable timeout.